File: equity_data.py
"""
Here is a list of Australian Companies by category:
http://www.asx.com.au/asx/research/companyInfo.do

Here is the format of Yahoo Finance downloads by category:
http://ichart.finance.yahoo.com/table.csv?d=6&e=1&f=2012&g=d&a=7&b=19&c=2003&ignore=.csv&s=ASX.AX

TODO:

* Think about foriegn exchanges' data, not just ASX
* what to do with missing data (so far, only companies that did not yet exist or folded)
* gzip data to save space
* return just a 2-column data set.
* how to represent dates? Ints, preferably.
"""

import requests
import csv
import os.path
import gzip
from settings import EQUITY_CACHE_DIR

YAHOO_URL_TEMPLATE = "http://ichart.finance.yahoo.com/table.csv?d=6&e=1&f=2012&g=d&a=1&b=1&c=2003&ignore=.csv&s=%s.AX"

FAVOURITE_FIRMS = \
{'1010': {
          'ORG': u'ORIGIN ENERGY',
          'RRS': u'RANGE RESOURCES LTD',
          'RUM': u'RUM JUNGLE RES',},
 '1510': {'AAI': u'ALCOA INC.',
          'BHP': u'BHP BILLITON LIMITED',
          'CYU': u'CHINALCO YUNNAN',
          'EAU': u'ELDORADO GOLD CORP',
          'FMG': u'FORTESCUE METALS GRP',
          'GDA': u'GONDWANA RESOURCES',
          'NCM': u'NEWCREST MINING',
          'ORD': u'ORD RIVER RESOURCES',
          'ORI': u'ORICA LIMITED',
          'OST': u'ONESTEEL LIMITED',
          'ZRI': u'RIO TINTO PLC',},
 '2010': {'CLO': u'CLOUGH LIMITED',
          'CSR': u'CSR LIMITED',},
 '2020': {'BXB': u'BRAMBLES LIMITED',
          'TSE': u'TRANSFIELD SERVICES',
          'ZBI': u'BRAMBLES IND. PLC'},
 '2030': {'AIZ': u'AIR NEW ZEALAND',
          'CLX': u'CTI LOGISTICS',
          'QAN': u'QANTAS AIRWAYS',
          'QRN': u'QR NATIONAL LIMITED',
          'SYD': u'SYD AIRPORT',},
 '2510': {'ION': u'ION LIMITED',
          'OEC': u'ORBITAL CORP LIMITED',},
 '2520': {'BBG': u'BILLABONG',
          'FPA': u'FISHER & PAYKEL APP.',},
 '2530': {'ALL': u'ARISTOCRAT LEISURE',
          'CWN': u'CROWN LIMITED',
          'DMP': u'DOMINO PIZZA ENTERPR',
          'FLT': u'FLIGHT CENTRE',
          'GEM': u'G8 EDUCATION LIMITED',
          'JET': u'JETSET TRAVELWORLD',
          'NVT': u'NAVITAS LIMITED',
          'SGH': u'SLATER & GORDON',
          'SKC': u'SKY CITY ENTERTAIN.',
          'TAH': u'TABCORP HOLDINGS LTD',
          'TTS': u'TATTS GROUP LTD',
          'VEL': u'VEALLS LIMITED',},
 '2540': {'AAU': u'ADCORP AUSTRALIA',
          'APN': u'APN NEWS & MEDIA',
          'CMJ': u'CONSOLIDATED MEDIA.',
          'FXJ': u'FAIRFAX MEDIA LTD',
          'NWS': u'NEWS CORP',
          'PNW': u'PACIFIC STAR NETWORK',
          'SKT': u'SKY NETWORK',
          'SWM': u'SEVEN WEST MEDIA LTD',
          'SXL': u'STHN CROSS MEDIA',
          'TEN': u'TEN NETWORK HOLDINGS',
          'VRL': u'VILLAGE ROADSHOW LTD',},
 '2550': {'CCV': u'CASH CONVERTERS',
          'CTY': u'COUNTRY ROAD LIMITED',
          'DJS': u'DAVID JONES LIMITED',
          'HVN': u'HARVEY NORMAN',
          'JBH': u'JB HI-FI LIMITED',
          'KMD': u'KATHMANDU HOLD LTD',
          'MYR': u'MYER HOLDINGS LTD',
          'PBG': u'PACIFIC BRANDS',
          'TRS': u'THE REJECT SHOP',
          'VTG': u'VITA GROUP LTD',
          'WEB': u'WEBJET LIMITED',
          'WTF': u'WOTIF.COM HOLDINGS'},
 '3010': {'WES': u'WESFARMERS LIMITED',
          'WOW': u'WOOLWORTHS LIMITED'},
 '3020': {'BGA': u'BEGA CHEESE LTD',
          'CCL': u'COCA-COLA AMATIL',
          'ELD': u'ELDERS LIMITED',
          'GFF': u'GOODMAN FIELDER.',},
 '3030': {'BKL': u'BLACKMORES LIMITED'},
 '3510': {'ANN': u'ANSELL LIMITED',
          'FPH': u'FISHER & PAYKEL H.',
          'OMI': u'OMI HOLDINGS LIMITED',
          'PHG': u'PULSE HEALTH LIMITED',
          'SIP': u'SIGMA PHARMACEUTICAL',
          'VEI': u'VISION EYE INSTITUTE'},
 '3520': {'AVH': u'AVITA MEDICAL LTD',
          'LCT': u'LIVING CELL TECH.',
          'TIS': u'TISSUE THERAPIES',
          'VSC': u'VITA LIFE SCIENCES.'},
 '4010': {'ANZ': u'ANZ BANKING GRP LTD',
          'BEN': u'BENDIGO AND ADELAIDE',
          'BOQ': u'BANK OF QUEENSLAND.',
          'CBA': u'COMMONWEALTH BANK.',
          'NAB': u'NATIONAL AUST. BANK',
          'WBC': u'WESTPAC BANKING CORP'},
 '4020': {'AAY': u'AACL HOLDINGS LTD',
          'AEF': u'AUSTRALIAN ETHICAL',
          'APD': u'APN PROPERTY GROUP',
          'APP': u'APA FINANCIAL',
          'AQF': u'AUS GOV INDEX FUND',
          'ARG': u'ARGO INVESTMENTS',
          'ASX': u'ASX LIMITED',
          'CNG': u'COLONIAL HOLDING LTD',
          'IFL': u'IOOF HOLDINGS LTD',
          'MQG': u'MACQUARIE GROUP LTD',
          'SOL': u'(SOUL PATTINSON (W.H))',},
 '4030': {'AMP': u'AMP LIMITED',
          'QBE': u'QBE INSURANCE GROUP',
          'SBK': u'SUNCORP-METWAY .',
          'SUN': u'SUNCORP GROUP LTD',},
 '4040': {'CER': u'CENTRO RETAIL GROUP',
          'CRF': u'CENTRO RETAIL AUST',
          'GMG': u'GOODMAN GROUP',
          'LMW': u'LANDMARK WHITE LTD',
          'MGI': u'MACQUARIE GOODMAN',
          'MGM': u'MACQUARIE GOOD. MGT.',
          'MGR': u'MIRVAC GROUP',
          'MIX': u'MIRVAC INDUSTRIAL',
          'MUE': u'MULTIPLEX EUROPEAN',
          'MXU': u'MULTIPLEX SITES',
          'SGP': u'STOCKLAND',
          'WDC': u'WESTFIELD GROUP',
          'WFA': u'WESTFIELD AMERICA',
          'WFT': u'WESTFIELD TRUST',
          'WRT': u'WESTFIELD RETAIL TST',
          'WSF': u'WESTFIELD HOLDINGS',
          'WWM': u'WENTWORTH HLDGS LTD'},
 '4510': {'IFM': u'INFOMEDIA LTD',},
 '4520': {'AAT': u'AAT CORPORATION LTD',},
 '5010': {'SGT': u'SINGAPORE TELECOMM.',
          'TEL': u'TELECOM CORPORATION',
          'TLS': u'TELSTRA CORPORATION.',
          'TPC': u'TEL.PACIFIC LIMITED',},
 '5510': {'AGK': u'AGL ENERGY LIMITED',
          'HRL': u'HOT ROCK LIMITED',}
}


def get_time_series(firm_code):
    """This is the iterator you usually wish to call"""
    with gzip.open(get_time_series_file(firm_code), 'rb') as f:
        reader = csv.reader(f, quoting=csv.QUOTE_NONE)
        for row in reader:
            yield row

def get_time_series_file(firm_code):
    cache_file_name = get_cache_file_name(firm_code)
    if not os.path.exists(cache_file_name):
        fetch_and_cache(firm_code)
    return cache_file_name

def get_cache_file_name(firm_code):
    return os.path.join(EQUITY_CACHE_DIR, firm_code + ".csv.gz")

def fetch_and_cache(firm_code):
    stock_data_request = requests.get(YAHOO_URL_TEMPLATE % firm_code)
    cache_file_name = get_cache_file_name(firm_code)
    with gzip.open(cache_file_name, 'wb') as cache_file:
        cache_file.write(stock_data_request.content)

def refresh_favourites_cache():
    for category, firm_dict in FAVOURITE_FIRMS.iteritems():
        for firm_code in firm_dict:
            fetch_and_cache(firm_code)

def refresh_all_cache():
    import asxinfo
    for category, firm_dict in asxinfo.ALL_INDUSTRY_FIRMS.iteritems():
        for firm_code in firm_dict:
            fetch_and_cache(firm_code)

File: equity_data.R
# Two granger libraries - we use the latter because it has missing value handling baked in,
# though the former is better for vectorised correlations
# Note that neither support true multivariate calculations, just pairwise.
#library("MSBVAR")
library("lmtest")
#library("ggplot2")
library("reshape2")
#library("ggdendro")
library("RSQLite")

# example usage:
# equities = pairwise.granger.test(get.favourite.indices(limit=Inf))
# #equities.m = pairwise.granger.test.m(get.favourite.indices(limit=Inf))


base.path = '/Users/dan/Dropbox/trade_data'
cache.path = paste(base.path, "cache", sep="/")

get.favourite.equities = function (limit=10) {
  files = list.files(cache.path, pattern=".*\\.csv\\.gz")
  favourite.equities.ticker.names = c('AAI', 'AAT', 'AAU', 'AAY', 'AEF', 'AGK', 'AIZ', 'ALL', 'AMP', 'ANN', 'ANZ', 'APD', 'APN', 'APP', 'AQF', 'ARG', 'ASX', 'AVH', 'BBG', 'BEN', 'BGA', 'BHP', 'BKL', 'BOQ', 'BXB', 'CBA', 'CCL', 'CCV', 'CER', 'CLO', 'CLX', 'CMJ', 'CNG', 'CRF', 'CSR', 'CTY', 'CWN', 'CYU', 'DJS', 'DMP', 'EAU', 'ELD', 'FLT', 'FMG', 'FPA', 'FPH', 'FXJ', 'GDA', 'GEM', 'GFF', 'GMG', 'HRL', 'HVN', 'IFL', 'IFM', 'ION', 'JBH', 'JET', 'KMD', 'LCT', 'LMW', 'MGI', 'MGM', 'MGR', 'MIX', 'MQG', 'MUE', 'MXU', 'MYR', 'NAB', 'NCM', 'NVT', 'NWS', 'OEC', 'OMI', 'ORD', 'ORG', 'ORI', 'OST', 'PBG', 'PHG', 'PNW', 'QAN', 'QBE', 'QRN', 'RRS', 'RUM', 'SBK', 'SGH', 'SGP', 'SGT', 'SIP', 'SKC', 'SKT', 'SOL', 'SUN', 'SWM', 'SXL', 'SYD', 'TAH', 'TEL', 'TEN', 'TIS', 'TLS', 'TPC', 'TRS', 'TSE', 'TTS', 'VEI', 'VEL', 'VRL', 'VSC', 'VTG', 'WBC', 'WDC', 'WEB', 'WES', 'WFA', 'WFT', 'WOW', 'WRT', 'WSF', 'WTF', 'WWM', 'ZBI', 'ZRI')
  i=0

  equities = NULL;

  for (file.name in files) {
    ticker.name = substr(file.name, 1, 3)
  
    if(!(ticker.name %in% favourite.equities.ticker.names)) {next}
    if(i>limit) {break}
    
    i=i+1
      
    one.equity = read.csv(gzfile(paste(cache.path, file.name, sep = "/")))
    trimmed.equity = one.equity[c("Date")]
    #take max because some shares still trade at 0 (granularity is $0.01, zero is -4.61)
    trimmed.equity[ticker.name] = log(pmax(
      one.equity["Adj.Close"],
      0.005))

    if(is.null(equities)) {
      equities = trimmed.equity
    } else {
      equities = merge(equities, trimmed.equity, all.x=TRUE, all.y=TRUE)
    }
  }
  equities$Date = as.Date(equities$Date)
  return(equities)
}

# This looks like it should go formula style.
# grangertest(AAT ~ AAY, data=equities)
# but constructing formulae in code is tedious,
# so we pass time series directly

get.ts = function(equities, ticker.name) {
  return(zoo(equities[,ticker.name], equities$Date))
}

granger.fp = function(xname, yname, data, order=1){
  res = NULL
  error = try(res <- grangertest(data[,xname], data[,yname], order=order))
  if(is.null(res)) return(list(f=NA,p=NA,worked=FALSE))
  f = res[2, "f"]
  p = res[2, "Pr(>F)"]
  return(list(f=f, p=p, worked=TRUE))
}
vec.granger.fp = Vectorize(granger.fp, vectorize.args=c('xname', 'yname'))

granger.p = function(xname, yname, data, order=1){
  return(as.numeric(granger.fp(xname, yname, data, order)['p']))
}
vec.granger.p = Vectorize(granger.p, vectorize.args=c('xname', 'yname'))

granger.f = function(xname, yname, data, order=1){
  return(as.numeric(granger.fp(xname, yname, data, order)['f']))
}
vec.granger.f = Vectorize(granger.f, vectorize.args=c('xname', 'yname'))

# returns a "melted" pairwise granger-causality distance frame
#Need to cast this to and from p/f matrices
#try:
# http://stackoverflow.com/a/9617424
# http://tolstoy.newcastle.edu.au/R/e6/help/09/01/0598.html
pairwise.granger.test = function(equities, order=1) {
  equity.names = names(equities)[-1]
  n = length(equity.names)
  n.pairs = n*(n-1)
  relations = data.frame
  sources = vector(mode='character', length = n.pairs)
  targets = vector(mode='character', length = n.pairs)
  fs = vector(mode='numeric', length = n.pairs)
  ps = vector(mode='numeric', length = n.pairs)
  i = 0
  
  for(j in 1:n) {
    for(k in (1:n)[-j]) {
      source.name = equity.names[j]
      target.name = equity.names[k]
      print(c(source.name, target.name))
      res = granger.fp(source.name, target.name, equities, order)
      i = i+1
      sources[i] = source.name
      targets[i] = target.name
      fs[i] = res$f
      ps[i] = res$p
    }
  }
  return(data.frame(source=as.factor(sources), target=as.factor(targets), f=fs, p=ps))
}

# returns a pairwise granger-causality distance matrix
# Need to cast this to and from sparse p/f frames
pairwise.granger.test.m = function(equities, order=1) {
  x = names(equities)[-1]
  names(x) = x
  y = x
  return(outer(x, y, vec.granger.f, equities))
}

#To see how to do this, try...
#  Traditional:
#    http://flowingdata.com/2010/01/21/how-to-make-a-heatmap-a-quick-and-easy-solution/
#    http://sphaerula.com/legacy/R/correlationPlot.html
#    http://www.phaget4.org/R/image_matrix.html
#  ggplot2:
#    https://learnr.wordpress.com/2010/01/26/ggplot2-quick-heatmap-plotting/
#    http://stackoverflow.com/a/5554352
#    http://stackoverflow.com/a/6675983 (bonus dendrogram!)
#    http://hosho.ees.hokudai.ac.jp/~kubo/Rdoc/library/ggmap/html/ggimage.html
#  general community structure:
#    http://sieste.wordpress.com/2012/05/21/inferring-the-community-structure-of-networks/
#    http://cran.cnr.berkeley.edu/web/views/Cluster.html
#
# linkcomm seems to do this for weighted digraphs.
# agnes (from cluster) doesn't like missing values, hclust (from stats) might be OK with 'em, but it isn't clear
# Here is an R GEXF gephi exporter (although a CSV export will probably do the trick)
# Drew conway chatter on the issue: http://www.drewconway.com/zia/?p=1221
# For a combo version, use the r heatmap plot with bonus dendrogram
#   heatmap(favourite.pairwise.vals)
#   more general clustering is in PDM, cluster, et al
# Sorting each axis separately might be informative enough without getting overexcited about directed graphs. 
# probably it can all go into SQL anyway

plot.correlation.matrix = function(correlations){
  
}

# For SQL, gephi likes nodes and edges in separate tables - 
# Edges:
# "source", "target", "label", "weight"
# Nodes:
# "id", "label", "x, "y", "size"
# For both of these we can add "start" and "end" to dynamic graphs
# here we respect the "source", "target" and "id" names but will use
# SELECT AS to coerce the remainder

# Convert a sparse pairwise correlation frame into a weighted, directed,
# SQL graph
# Gephi can interpret this using the folloing nodes/edges queries respectively
# SELECT id, id AS "label" FROM nodes
# SELECT source, target, f AS "weight" FROM edges
correlations.to.sql = function(data, dbname="equities_graph.db", max.p=0.05) {
  dbpath = paste(base.path, dbname, sep="/")
  print(c("opening", dbpath))
  conn <- dbConnect("SQLite", dbname = dbpath)
  
  data = data[!is.na(data$p),]
  data = data[data$p<max.p,]
  node.names = data.frame(id=unique(data$source))
  
  dbWriteTable(conn, "nodes", node.names)
  dbWriteTable(conn, "edges", data)
  dbDisconnect(conn)
}

# Convert a weighted, directed, SQL graph into a sparse pairwise
# correlation frame
sql.to.correlations = function(dbname="equities_graph.db",
      q = "SELECT * from edges") {
  dbpath = paste(base.path, dbname, sep="/")
  print(c("opening", dbpath))
  conn <- dbConnect("SQLite", dbname = dbpath)
  res = dbGetQuery(conn, q)
  dbDisconnect(conn)
  return(res)
}